src/super_context.ml

```diff
@@ -287,23 +287,16 @@ let create
   let installed_libs =
     Lib.DB.create_from_findlib context.findlib ~external_lib_deps_mode
   in
-  let internal_libs =
-    List.concat_map stanzas
-      ~f:(fun { Dune_load.Dune_file. dir; stanzas; project = _ ; kind = _ } ->
-        let ctx_dir = Path.append context.build_dir dir in
-        List.filter_map stanzas ~f:(fun stanza ->
-          match (stanza : Stanza.t) with
-          | Library lib -> Some (ctx_dir, lib)
-          | _ -> None))
-  in
-  let internal_coq_libs =
-    List.concat_map stanzas
+  let internal_libs, internal_coq_libs =
+    List.map stanzas
       ~f:(fun { Dune_load.Dune_file. dir; stanzas; project = _ ; kind = _ } ->
         let ctx_dir = Path.append context.build_dir dir in
-        List.filter_map stanzas ~f:(fun stanza ->
+        List.filter_partition_map stanzas ~f:(fun stanza ->
           match (stanza : Stanza.t) with
-          | Coq.T coq_lib -> Some (ctx_dir, coq_lib)
-          | _ -> None))
+          | Library lib -> Left (ctx_dir, lib)
+          | Coq.T coq_lib -> Right (ctx_dir, coq_lib)
+          | _ -> Skip))
+    |> List.split |> (fun (ll,lr) -> List.concat ll, List.concat lr)
```

Comment:
If this is about improving performance, then I think most of the gains will be lost by the subsequent split and concatenations. I suggest to just write this with folds if you think it's important to make this fast. It could gain some clarity as well I think.

Reply:
I dunno what to do here, the original code was using `concat_map` which is also quite unoptimal; more than performance I was worried about keeping a "single pass" over the stanzas, somehow it should be a bit more generic than we have now indeed; I'll give it a go.

Reply:
I just replaced the code with two folds. See the commit.